don't use "role" or "roles" to include roles

rishabh-d-dave opened the pull request:
Use `import_task` and `include_task` instead.

leseb commented:
@andrewschoen can you help with the failure here: https://2.jenkins.ceph.com/job/ceph-ansible-prs-luminous-update_cluster/49/consoleFull#-729725784b5dd38fa-a56e-4233-a5ca-584604e56e3a

```
TASK [ceph-validate : validate provided configuration] *************************
task path: /home/jenkins-build/build/workspace/ceph-ansible-prs-luminous-update_cluster/roles/ceph-validate/tasks/main.yml:2
Thursday 23 August 2018  05:09:09 +0000 (0:01:21.304)       0:03:18.661 ******* 
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'containerized_deployment'
fatal: [ceph-mon0]: FAILED! => {}

MSG:

Unexpected failure during module execution.
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'containerized_deployment'
fatal: [ceph-mon1]: FAILED! => {}

MSG:

Unexpected failure during module execution.
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: KeyError: 'containerized_deployment'
fatal: [ceph-osd0]: FAILED! => {}
...
...
...
```

Thanks

rishabh-d-dave commented:
Most CI jobs had failed because `ceph_origin` went through the problem I mentioned in my [last comment](https://github.com/ceph/ceph-ansible/pull/3058#issuecomment-418056075): with my changes, `ceph_origin` is undefined which is not the case otherwise.

rishabh-d-dave commented:
Rebasing the PR.

andrewschoen commented on site.yml.sample:
```diff
@@ -72,30 +72,35 @@
-    - ceph-defaults
-    - ceph-validate
-
+    - include_vars: roles/ceph-defaults/defaults/main.yml
```
I believe this will override any variables we've set in other places as vars included in a task take precedence over vars set in the inventory or a group_vars file. This might be your problem with ``ceph_origin``.  You can see from the CI log (https://2.jenkins.ceph.com/job/ceph-ansible-prs-luminous-update_cluster/52/consoleFull) that ``ceph_origin`` is being reset to ``dummy``.

```
TASK [include_vars] ************************************************************
task path: /home/jenkins-build/build/workspace/ceph-ansible-prs-luminous-update_cluster/site.yml.sample:75
Wednesday 05 September 2018  09:33:05 +0000 (0:00:00.409)       0:00:45.308 *** 
ok: [ceph-mon0] => {
    "ansible_facts": {
        "bootstrap_dirs_group": "64045", 
        "bootstrap_dirs_owner": "64045", 
        "centos_package_dependencies": [
            "python-pycurl", 
            "epel-release", 
            "python-setuptools", 
            "libselinux-python"
        ], 
        "ceph_conf_key_directory": "/etc/ceph", 
        "ceph_conf_local": false, 
        "ceph_conf_overrides": {}, 
        "ceph_custom_repo": "https://server.domain.com/ceph-custom-repo", 
        "ceph_dev_branch": "master", 
        "ceph_dev_sha1": "latest", 
        "ceph_docker_enable_centos_extra_repo": false, 
        "ceph_docker_image": "ceph/daemon", 
        "ceph_docker_image_tag": "latest", 
        "ceph_docker_on_openstack": false, 
        "ceph_docker_registry": "docker.io", 
        "ceph_iscsi_config_dev": true, 
        "ceph_iscsi_firewall_zone": "public", 
        "ceph_keyring_permissions": "0600", 
        "ceph_mds_firewall_zone": "public", 
        "ceph_mgr_firewall_zone": "public", 
        "ceph_mirror": "http://download.ceph.com", 
        "ceph_mon_firewall_zone": "public", 
        "ceph_nfs_firewall_zone": "public", 
        "ceph_origin": "dummy", 
        "ceph_osd_firewall_zone": "public", 
        "ceph_rbdmirror_firewall_zone": "public",
 ...
```

rishabh-d-dave replied:
Agreed. I added it for debugging and later forgot to get rid of it.

rishabh-d-dave commented:
retest this please

rishabh-d-dave commented:
* Added `private: false` to all imports and includes tasks since we are using the default behaviour of making variable visible in the play.
* Removed PR #3016's commit.
* Added a commit message.
* Added `skip ci` to the PR title since this PR depends on the Ansible release 2.7 (which would added the functionality of making variables visible even when `import_role` and `include_role` is used).

*Note*: 2.7 releases by [4 Oct 2018](https://docs.ansible.com/ansible/devel/roadmap/ROADMAP_2_7.html)

guits commented:
jenkins test pipeline

leseb commented:
@guits retargeted for 3.3 then

rishabh-d-dave commented:
retest this please

rishabh-d-dave commented:
@guits @leseb 
The reason for failure of update_cluster is that `bluestore_wal_devices` is defined using a Jinja2 template which is being passed un-processed/expanded/substituted to validate plugin's code. So, IMO, the CI failure can be classified as not relevant to this PR.

rishabh-d-dave commented:
@guits @leseb 

Rishabh Dave [wrote](https://github.com/ceph/ceph-ansible/pull/3058#issuecomment-429331749):
> @guits @leseb
> The reason for failure of update_cluster is that bluestore_wal_devices is defined using a Jinja2 template which is being passed un-processed/expanded/substituted to validate plugin's code. So, IMO, the CI failure can be classified as not relevant to this PR.

The workaround worked -- update_cluster job was completed successfully!

rishabh-d-dave commented:
jenkins test luminous-switch_to_containers

rishabh-d-dave commented:
@leseb 
The [test-pipeline](https://2.jenkins.ceph.com/job/ceph-ansible-pipeline/1484/) failed at the job [luminous-switch_to_containers](https://2.jenkins.ceph.com/job/ceph-ansible-prs-luminous-switch_to_containers/35/), However, it passed later when I [triggered it individually](https://2.jenkins.ceph.com/job/ceph-ansible-prs-luminous-switch_to_containers/36/).

Since there is no way to resume running the test pipeline from the failing job that stopped it, is there way to know which jobs would have ran next given the failing jobs would have passed? If I could just run that bunch of jobs, it would save CI resources, avoid repetition and I could know the results of CI faster.

leseb commented:
@rishabh-d-dave, please rebase due to https://github.com/ceph/ceph-ansible/pull/3175, thanks. We are getting close.

guits commented on plugins/actions/validate.py:
```diff
@@ -28,6 +29,15 @@ class ActionModule(ActionBase):
         host_vars = task_vars['vars']
+
+        # process jinja2 templates
+        reqd_type = "<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>"
```
@andrewschoen do you mind taking a look at this, please?

guits replied:
@rishabh-d-dave ok, i've just noticed your email

rishabh-d-dave commented:
jenkins test pipeline

rishabh-d-dave commented:
jenkins test pipeline

rishabh-d-dave commented:
jenkins test pipeline

rishabh-d-dave commented:
jenkins retest this please

rishabh-d-dave commented:
@guits @leseb 
This PR consistently fails on [ceph-ansible-prs-luminous-add_osds_container](https://2.jenkins.ceph.com/job/ceph-ansible-prs-luminous-add_osds_container/14/). What CI jobs should I trigger to test this PR?

rishabh-d-dave commented:
jenkins test luminous-docker_cluster

rishabh-d-dave commented:
jenkins test luminous-centos7_cluster

rishabh-d-dave commented:
jenkins test luminous-ooo_collocation

rishabh-d-dave commented:
jenkins test luminous-centos7_cluster

rishabh-d-dave commented:
@leseb 
All the manually triggered tests passed.

rishabh-d-dave commented:
@leseb 
Should I get rid of DNM commits on this PR? Doing so would fail the CI since one of them is a fix for issue [#3247](https://github.com/ceph/ceph-ansible/issues/3247) and other helps this PR test against Ansible 2.7.

leseb commented:
@rishabh-d-dave please rebase

rishabh-d-dave commented:
jenkins test luminous-centos7_cluster

rishabh-d-dave commented:
jenkins test luminous-docker_cluster

rishabh-d-dave commented:
jenkins test luminous-ooo_collocation

alfredodeza commented:
> Use `import_task` and `include_task` instead.

Why?

alfredodeza commented:
@guits thanks! the commit message does explain it. We've faced a bunch of issues related to this PR, and I wasn't able to parse the need. I guess the reasoning was to make it more readable and explicit

dmick commented:
I spent three days tracking down why this broke my installation.  It requires Ansible 2.7, see https://github.com/ansible/ansible/pull/41330; otherwise, any defaults set in ceph-defaults are not available to the calling play.

Also, the ceph-validate code is broken about trying to inform users about why this happens; see fixes to notario (https://github.com/alfredodeza/notario/pull/19 and https://github.com/alfredodeza/notario/pull/20) and upcoming fixes to validate.py to print usable error messages rather than triggering tracebacks.

dmick commented:
https://github.com/ceph/ceph-ansible/pull/3336 for validate.py changes
